packages/lazy-import/lib/index.js

```diff
@@ -0,0 +1,114 @@
+ */
+
+/**
+ * Returns an md5 hash of the given string.
+ *
+ * @param {string} text Text for which to generate hash.
+ *
+ * @return {string} md5 hash of string.
+ */
+function md5( text ) {
+	return createHash( 'md5' )
+		.update( text )
+		.digest( 'hex' );
+}
+
+/**
+ * Given an arg string as would be passed to `npm install`, returns a string to
+ * use as the alias name of the package to install locally if needed.
+ *
+ * @param {string} arg A string that you might pass to `npm install`.
+ *
+ * @return {string} Module to use as local package alias.
+ */
+function getLocalModuleName( arg ) {
+	return '@wordpress/lazy-import.' + md5( arg );
+}
+
+/**
+ * Installs npm package by arg name at the given prefix path. Creates the prefix
+ * path if it doesn't yet exist.
+ *
+ * @param {string} arg   Package install arg (name and version specification).
+ * @param {string} alias Alias to use for local module name.
+ *
+ * @return {Promise<void>} Promise resolving once package is installed.
+ */
+async function install( arg, alias ) {
+	await execa( 'npm', [ 'install', '--no-save', alias + '@npm:' + arg ] );
+}
+
+/**
+ * Given an arg string as would be passed to `npm install`, requires the package
+ * corresponding to the name and specifier parsed from the arg. If the package
+ * is not installed (or not installed at that version), the package is installed
+ * to a temporary directory at `node_modules/.wp-lazy` relative to where the
+ * current working directory.
+ *
+ * @param {string}                       arg       A string that you might pass
+ *                                                 to `npm install`.
+ * @param {Partial<WPLazyImportOptions>} [options] Optional options object.
+ *
+ * @return {Promise<NodeRequire>} Promise resolving to required module.
+ */
+async function lazyImport( arg, options = {} ) {
+	const { rawSpec, name } = npa( arg );
+
+	if ( ! name ) {
+		throw new TypeError(
+			`Unable to parse package name from \`${ arg }\`.`
+		);
+	}
+
+	const localModule = getLocalModuleName( arg );
+
+	// Try first from the temporary install path, since the second attempt will
+	// need to verify both availability and version. Version isn't necessary to
+	// account for in this first attempt.
+	try {
+		return require( localModule );
+	} catch ( error ) {
+		if ( error.code !== 'MODULE_NOT_FOUND' ) {
+			throw error;
+		}
+	}
+
+	try {
+		const resolved = require( name );
+
+		const { version } = require( join( name, 'package.json' ) );
+		if ( semver.satisfies( version, rawSpec ) ) {
+			// Only return with the resolved module if the version is valid per
+			// the parsed arg. Otherwise, fall through to install stage.
+			return resolved;
+		}
+	} catch ( error ) {
+		if ( error.code !== 'MODULE_NOT_FOUND' ) {
+			throw error;
+		}
+	}
+
+	// If this point is reached, the module cannot be found and must be
+	// installed.
+	if ( options.onInstall ) {
+		options.onInstall();
+	}
+	await install( arg, localModule );
```

Comment:
Assuming a user is offline, this will fall. Does it require a special handling?

Reply:
> Assuming a user is offline, this will fall. Does it require a special handling?

Good points. It's a similar dilemma to what we faced with https://github.com/WordPress/gutenberg/pull/21306#issuecomment-607281261, in that: it's not exactly ideal that a developer would be prevented from proceeding due to the fact that a lazily-installed dependency can't be installed due to being offline. In some cases (like what's implemented here with benchmarking) it might be okay. Others might need more consideration, such as if there's desire to use it with `wp-scripts` commands.

I don't know that there's anything special we could do at the package level, assuming we want to leave it generic enough to not assume anything specific about the context in which this is being run (e.g. a terminal). I think the alternatives could be to throw an error (reject the promise), or to add an extra option similar to what's available with `onInstall`.

Error seems like the most obvious candidate, though I think it'd require some additional documentation to help a consumer anticipate this scenario.

Reply:
> Assuming a user is offline, this will fall. Does it require a special handling?

I added some special handling and documentation for this in d883925427.